Add rich comparison operators to FunctionxD

jacklovell opened the pull request:
Enable arithmetic comparison operations between FunctionxD objects,
which return 1.0 if the comparison returns True, or 0.0 if the
comparison returns false.

At the moment I've only implemented the 2D operators, so we can discuss the implementation before I go ahead and add the 1D and 3D ones. The pull request would end up being too large to properly review otherwise.

Also, should these features still be based off the feature/function branch? Or is the function framework sufficiently stable that we can add new functions straight to the development branch?

CnlPepper commented:
Thanks Jack, I only just noticed the email about this merge request. I'll have look through asap.

jacklovell commented:
No worries. I won't be doing any more work on this until #326 is merged, then I'll rebase off of that. I really only want to know whether you agree with the approach taken here.

jacklovell commented:
@CnlPepper I've updated this PR with the new wrapping behaviour, and also addressed all the other comments I remember from our discussion on Wednesday. This is now ready for a preliminary review, and if you're happy with it then I'll add the 1D and 3D cases.

CnlPepper commented:
Hi Jack, sorry for the delay. I'll have a look through later this afternoon/evening if I can.

jacklovell commented:
No worries. I've just rebased this onto the latest development (with sqrt and erf), hence the force push. No other changes yet.

CnlPepper commented on raysect/core/math/function/function2d/base.pyx:
```diff
@@ -162,6 +163,39 @@ cdef class Function2D:
+    def __richcmp__(self, object other, int op):
+        if is_callable(other):
+            if op == Py_EQ:
+                return EqualsFunction2D(self, <Function2D> other)
```
Remove the Function2D cast here, otherwise pure python functions will fail. The autowrap in the init will handle the conversion.

jacklovell replied:
Thanks, I missed these. I'll remove the casts for all `op`s.

CnlPepper commented on raysect/core/math/function/function2d/tests/test_base.py:
```diff
@@ -137,6 +137,126 @@ def test_pow_function2d_scalar(self):
+        for x in v:
+            for y in v:
+                ref_value = self.ref1(x, y)
+                greater_value = ref_value + abs(ref_value) + 1
```
Being picky here, but could you rename these to higher_value and lower_value as it is easier to interpret.

jacklovell replied:
Yep, I can do this

CnlPepper commented on raysect/core/math/function/function2d/tests/test_base.py:
```diff
@@ -237,4 +357,201 @@ def test_pow_3_arguments(self):
+        for x in v:
+            for y in v:
+                ref_value = self.ref1
+                greater_value = lambda x, y: self.ref1(x, y) + abs(self.ref1(x, y)) + 1
```
Same rename as above.

jacklovell replied:
Fine

CnlPepper commented on raysect/core/math/function/function2d/tests/test_base.py:
```diff
@@ -237,4 +357,201 @@ def test_pow_3_arguments(self):
+                # Without Function2D.__abs__, manually ensure greater and less values are suitable
+                if self.f1(x, y) >= 0:
+                    greater_value = self.f1 + self.f1 + 1
+                    less_value = self.f1 - self.f1 - 1
```
isn't this just -1? Is this what was intended?

jacklovell replied:
I remember why I did this now. This is `Constant2D(-1)`, not `-1`. The tests are `function __richcmp__ function`, so I need a `Function2D` object here rather than a scalar for the comparisons. The expression comes from a copy/paste/tweak of the function-scalar test, working around the lack of `__abs__`.

I can either add a comment explaining this, or explicitly use `Constant2D(-1)` instead. Or implement `__abs__` and make this PR even bigger. Which would you prefer?

CnlPepper replied:
Adding abs would be preferable.

jacklovell replied:
OK, I've added abs and rewritten the test to use it. If you're happy with the abs implementation, I'll add 1D and 3D versions of `__richcmp__` and `__abs__` and then mark this PR as ready for merging.

CnlPepper commented on raysect/core/math/function/function2d/tests/test_base.py:
```diff
@@ -237,4 +357,201 @@ def test_pow_3_arguments(self):
+                    greater_value = self.f1 + self.f1 + 1
+                    less_value = self.f1 - self.f1 - 1
+                else:
+                    greater_value = self.f1 - self.f1 + 1
```
isn't this just 1?  Is this what was intended?

jacklovell replied:
See above

jacklovell replied:
Resolved by adding abs operator.

CnlPepper commented:
Ah you planning to add the 1D and 3D implementations in this merge request or will these follow in a separate request?

jacklovell commented:
I was planning to add the 1D and 3D implementations in this merge request on Monday morning. But if you want to merge this and add them yourself before then don't let me stop you!

CnlPepper commented:
He he, oh I'm happy to wait for Monday. :)

jacklovell commented:
@CnlPepper I've added the 1D and 3D cases. This is now ready for review and merge.

CnlPepper commented:
Thanks @jacklovell
